features/430-passwordless/design.md

```diff
@@ -0,0 +1,204 @@
+    - Time-based OTP: TOTP parameters are generated and given to client,
+                      and a valid TOTP must be entered to activate it.
+    - OOB OTP: A OTP is sent to the OOB channel, and the received OTP must be
+               entered to activate it.
+    - WebAuthn: TBD.
+- Developer can configure multiple authenticators:
+    - At least one primary authenticators must be configured, and secondary
+      authenticator is optional.
+    - Authenticator types can be configured as either primary or secondary.
+    - During authentication, one of the configured primary authenticators is
+      needed, and one of the configured secondary authenticators is needed.
+    - For examples:
+        - Simple password/SSO login:
+            - primary: password/SSO.
+            - no secondary.
+        - Password/SSO with email/TOTP MFA:
+            - primary: password/SSO.
+            - secondary: OOB-OTP/TOTP/Recovery Code.
+        - WebAuthn password-less with resident key:
+            - primary: WebAuthn.
+            - no secondary.
+        - Email OTP passwordless/SSO:
+            - primary: OOB-OTP/SSO.
+            - no secondary.
+        - GitHub:
+            - primary: password.
+            - secondary: OOB-OTP/TOTP/Recovery Code/WebAuthn.
+- Developer can require authenticators to be set up.
+    - A primary authenticator must be set up in the sign-up process.
+    - Secondary authenticator is set up after authentication.
+        - If secondary authentication is configured as required,
+          and no secondary authenticator is set up, user must setup secondary
+          authenticators before a session is created.
+        - If secondary authentication is configured as optional,
+          user can setup secondary authenticators in Auth UI settings page.
+
+## Authentication session
+
+The current authentication session goes like this:
+1. Authenticate Identity
+2. Perform MFA/Setup MFA
+
+Since selecting and authenticating the identity is separated under the new
+model, there will be extra steps in the authentication session:
+1. Select Identity
+2. Perform Primary Authentication
+    - If multiple primary authenticators is applicable for the selected
+      identity, user is prompted to select one.
+3. Perform Secondary Authentication (if configured)
+    - If multiple secondary authenticators is set up for the user, user is
+      prompted to select one.
+    - If no secondary authenticators is configured:
+        - If secondary authentication is required, user is prompted to
+          setup a secondary authenticator.
+        - If secondary authentication is optional, this step is skipped.
+
+For examples:
+- Simple password login:
+    1. Enter Login ID.
+    2. Authenticate with password.
+    3. Done.
+- Password/SSO with email/TOTP MFA:
+    1. Select Login ID or OAuth identity.
+        - Login ID: enter email.
+        - OAuth: redirect to OAuth provider and wait for return.
```

Comment:
Ben in the past told me a feature like this. After entering the email, if the email matches one of the OAuth identity the user has, redirect the user to the OAuth provider. So what the user really enters is a hint which can be interpreted as login ID or a reference to one of the OAuth identity they have. Anyway it is an improvement and (I think) no need to be done at this stage.

Reply:
Should primary authenticator have priority? A recurring requirement I have been told is that the user initially signed up with `user@gmail.com`, later on the user connected their gmail account. From now on the user must use google to login and cannot use password anymore.

Reply:
> After entering the email, if the email matches one of the OAuth identity the user has, redirect the user to the OAuth provider.

Yes, it is intended to be possible to support that. A more generalized formalization is that, in this step user gives a set of claims (e.g. `{email: user@example.com}` for email login ID, `{oauth_provider: google}` for OAuth), then the system would match the claims to find a candidate identity, and it is up to the frontend to interpret user input into this set of claims.

We definitely do not need this right now.

> A recurring requirement I have been told is that the user initially signed up with user@gmail.com, later on the user connected their gmail account. From now on the user must use google to login and cannot use password anymore.

We should not have this problem, since user can continue to enter login ID and password to login, as long as the login ID identity and password authenticator is intact, and system is configured to allow these identity and authenticator types to be used.

Reply:
Sorry, I meant the requirement is:

As long as the user has an OAuth Identity whose claims match one of their login ID, then the login ID must not be used anymore. Further login must use the OAuth identity.

Reply:
If so, I think using priority is not appropriate in this scenario: the use case wants to 'forbid' login ID identity to be used, not 'prioritize' OAuth identity. Misusing semantics of features can lead to future breaks.

For this use case, I'll suggest the developer to delete the login ID identity/password authenticator in post create hook of OAuth identity.

Reply:
For the hook approach to work, the developer need to has access to Admin API that allows them to query additional information about the user. For example in `after_identity_create`, the developer needs to query all identities the user, to see if the identities match the condition.

Reply:
Yes, that's correct. The simplest approach is to get the list of identities and delete login ID identity, if the created identity is an OAuth identity. We're going to have proper admin APIs right?

Reply:
> We're going to have proper admin APIs right?

Yes

---

Just got this idea in my mind. We could have some implemented hooks to solve some common problems. The developer can just enable the hooks in config.

This feature and the OnUserDuplicate flag could be implemented by hook.

Reply:
> Just got this idea in my mind. We could have some implemented hooks to solve some common problems. The developer can just enable the hooks in config.
> 
> This feature and the OnUserDuplicate flag could be implemented by hook.

I'm not sure if that's a good idea:
- Hooks are supposed to be extension point for external services, and handling hook events by ourselves seems strange.
- Once implemented, we cannot have breaking changes to the behavior of built-in hooks.
- If there is a bug in built-in hook handlers, either user should update to latest version with the fix, which may have undesired changes, or we back-port the fix to previous versions, and neither one seems ideal.
- It would tend to be misused, if it is the only endpoint exposing certain behaviors.
- We need to have a clear criteria on what problems are 'common' enough to have the hooks included in-repo, since everybody would certainly want us to maintain the hooks.